File: src/Scripting/Perl/interpreter.pl
use strict;
use warnings;

sub number {
    my ($value) = @_;
    return { kind => 'number', value => $value };
}

sub add {
    my ($left, $right) = @_;
    return { kind => 'add', left => $left, right => $right };
}

sub interpret {
    my ($expr) = @_;
    return $expr->{value} if $expr->{kind} eq 'number';
    return interpret($expr->{left}) + interpret($expr->{right}) if $expr->{kind} eq 'add';
    die "unknown expression\n";
}

my $expression = add(add(number(2), number(3)), number(4));
my $value = interpret($expression);
die "unexpected value\n" unless $value == 9;
print "interpreter=$value\n";

File: src/Systems/Go/interpreter.go
package main

import "fmt"

// Interpreter in idiomatic Go: grammar nodes are values that recursively
// interpret themselves. The tiny grammar is Expr := Number | Expr + Expr.
type Expr interface {
	Interpret() int
}

type Number struct {
	Value int
}

func (n Number) Interpret() int {
	return n.Value
}

type Add struct {
	Left  Expr
	Right Expr
}

func (a Add) Interpret() int {
	return a.Left.Interpret() + a.Right.Interpret()
}

func main() {
	expr := Add{
		Left: Number{Value: 2},
		Right: Add{
			Left:  Number{Value: 3},
			Right: Number{Value: 4},
		},
	}

	value := expr.Interpret()
	if value != 9 {
		panic(fmt.Sprintf("expected 9, got %d", value))
	}

	fmt.Printf("value=%d\n", value)
}
